File: src/utils/xmlBuilder.js
/**
 * ============================================================
 * FIL: src/utils/xmlBuilder.js
 * FORMÅL: Bygger en SOAP-lignende XML-melding fra en POS-ordre.
 *
 * HVA ER SOAP OG XML?
 *  - XML (eXtensible Markup Language) er et format for å strukturere data,
 *    litt som JSON men med åpne og lukkede tagger (som HTML).
 *  - SOAP er en protokoll (et sett regler) for å sende meldinger
 *    mellom systemer. SOAP bruker alltid XML som format.
 *  - Mange eldre systemer (inkl. mange PC-kasser) bruker SOAP
 *    fordi det er veldig strengt definert og forutsigbart.
 *
 * EKSEMPEL PÅ XML vs JSON:
 *  JSON:  { "navn": "Marcus" }
 *  XML:   <Navn>Marcus</Navn>
 *
 * I VIRKELIGHETEN ville denne XML-en bli sendt over HTTP til
 * et SOAP-endepunkt – men her skriver vi den bare ut til konsoll.
 * ============================================================
 */

/**
 * byggProduktXml
 * --------------
 * Hjelpefunksjon som lager XML for ett enkelt produkt.
 * Kalles én gang per produkt i ordren.
 *
 * @param {object} produkt - Et produkt-objekt fra POS-ordren
 * @returns {string} - XML-streng for produktet
 */
function byggProduktXml(produkt) {
  return `
        <Produkt>
            <Produktnavn>${produkt.produktnavn}</Produktnavn>
            <SKU>${produkt.sku}</SKU>
            <Antall>${produkt.antall}</Antall>
            <Enhetspris>${produkt.enhetspris}</Enhetspris>
            <Linjetotal>${produkt.linjetotal}</Linjetotal>
        </Produkt>`;
}

/**
 * byggSoapXml
 * -----------
 * Bygger en komplett SOAP-envelope (konvolutt) med ordre-data.
 *
 * SOAP-struktur (forenklet):
 *  <Envelope>        ← Ytterste wrapper – alltid påkrevd i SOAP
 *    <Body>          ← Innholdet i meldingen
 *      <OrdreRequest> ← Vår egne data
 *        ...
 *      </OrdreRequest>
 *    </Body>
 *  </Envelope>
 *
 * @param {object} posOrdre - POS-ordre-objekt fra mapperen
 * @returns {string} - Komplett SOAP XML-streng
 */
function byggSoapXml(posOrdre) {
  // Bygg XML for alle produkter og slå dem sammen til én streng
  // .map() lager array av XML-strenger, .join("") slår dem sammen
  const produkterXml = posOrdre.produkter.map(byggProduktXml).join("");

  // Template literal (backticks) gjør det enkelt å bygge lange strenger
  // med variabler inni
  return `<?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
    <soapenv:Body>
        <OrdreRequest>
            <OrdreId>${posOrdre.ordreId}</OrdreId>
            <Kundenavn>${posOrdre.kundenavn}</Kundenavn>
            <KundeEpost>${posOrdre.kundeEpost}</KundeEpost>
            <FraktMetode>${posOrdre.fraktMetode}</FraktMetode>
            <FraktPris>${posOrdre.fraktPris}</FraktPris>
            <TotalBelop>${posOrdre.totalBelop}</TotalBelop>
            <Produkter>${produkterXml}
            </Produkter>
        </OrdreRequest>
    </soapenv:Body>
</soapenv:Envelope>`;
}

// ---------------------------------------------------------------
// EKSPORT
// ---------------------------------------------------------------
module.exports = { byggSoapXml };

/**
 * ============================================================
 * OPPSUMMERING:
 *  - byggProduktXml(): lager XML for ett produkt (intern hjelpefunksjon)
 *  - byggSoapXml(): lager komplett SOAP-envelope med hele ordren
 *  - XML-strukturen etterligner hva et ekte SOAP-kall ville sendt
 *  - I .NET ville vi brukt en WSDL og auto-genererte klasser,
 *    men prinsippet er det samme: strukturert XML med faste feltnavn
 * ============================================================
 */

File: dotnet/ShopifyPosIntegrasjon/Utils/XmlBuilder.cs
// ============================================================
// FIL: Utils/XmlBuilder.cs
// FORMÅL: Bygger SOAP XML-melding fra en POS-ordre.
//         Tilsvarer xmlBuilder.js i Node.js-versjonen.
//
// SAMMENLIGNING MED JAVASCRIPT:
//  JS:  posOrdre.produkter.map(byggProduktXml).join("")
//  C#:  string.Concat(posOrdre.Produkter.Select(ByggProduktXml))
//
//  JS:  Template literals (backticks): `<Tag>${verdi}</Tag>`
//  C#:  String interpolation: $"<Tag>{verdi}</Tag>"
//
// I .NET PRODUKSJON ville vi brukt:
//  - System.Xml.Linq (XDocument/XElement) for strukturert XML-bygging
//  - Eller auto-genererte klasser fra en WSDL-fil via "Add Service Reference"
//  Her bruker vi string-bygging for å speile JS-versjonen direkte.
// ============================================================

namespace ShopifyPosIntegrasjon.Utils;

public class XmlBuilder
{
    // Tilsvarer "function byggSoapXml(posOrdre)" i JS
    public string ByggSoapXml(PosOrder posOrdre)
    {
        // string.Concat() tilsvarer .map(...).join("") i JS
        string produkterXml   = string.Concat(posOrdre.Produkter.Select(ByggProduktXml));
        string fraktLinjerXml = string.Concat(posOrdre.FraktLinjer.Select(ByggFraktXml));

        // Bygg rabatt-XML kun hvis det finnes rabatter
        string rabatterXml = posOrdre.Rabatter.Count > 0
            ? "<Rabatter>" + string.Concat(posOrdre.Rabatter.Select(ByggRabattXml)) + "\n            </Rabatter>"
            : "";

        // @"..." er en verbatim string – linjeskift og innrykk bevares
        // Tilsvarer template literals (backticks) i JS
        return
            $"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
            $"<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\">\n" +
            $"    <soapenv:Body>\n" +
            $"        <OrdreRequest>\n" +
            $"            <OrdreId>{posOrdre.OrdreId}</OrdreId>\n" +
            $"            <Kundenavn>{posOrdre.Kundenavn}</Kundenavn>\n" +
            $"            <KundeEpost>{posOrdre.KundeEpost}</KundeEpost>\n" +
            $"            <Fraktlinjer>{fraktLinjerXml}\n" +
            $"            </Fraktlinjer>\n" +
            $"            <TotalFrakt>{posOrdre.TotalFrakt}</TotalFrakt>\n" +
            $"            {rabatterXml}\n" +
            $"            <TotalRabatt>{posOrdre.TotalRabatt}</TotalRabatt>\n" +
            $"            <TotalBelop>{posOrdre.TotalBelop}</TotalBelop>\n" +
            $"            <Produkter>{produkterXml}\n" +
            $"            </Produkter>\n" +
            $"        </OrdreRequest>\n" +
            $"    </soapenv:Body>\n" +
            $"</soapenv:Envelope>";
    }

    // "private static" = intern hjelpemetode, ingen instans nødvendig
    // Tilsvarer "function byggProduktXml(produkt)" i JS
    private static string ByggProduktXml(PosProdukt produkt) =>
        $"\n        <Produkt>" +
        $"\n            <Produktnavn>{produkt.Produktnavn}</Produktnavn>" +
        $"\n            <SKU>{produkt.Sku}</SKU>" +
        $"\n            <Antall>{produkt.Antall}</Antall>" +
        $"\n            <Enhetspris>{produkt.Enhetspris}</Enhetspris>" +
        $"\n            <Linjetotal>{produkt.Linjetotal}</Linjetotal>" +
        $"\n        </Produkt>";

    private static string ByggFraktXml(PosFraktlinje linje) =>
        $"\n        <Fraktlinje>" +
        $"\n            <Metode>{linje.Metode}</Metode>" +
        $"\n            <Pris>{linje.Pris}</Pris>" +
        $"\n        </Fraktlinje>";

    private static string ByggRabattXml(PosRabatt rabatt) =>
        $"\n        <Rabatt>" +
        $"\n            <Kode>{rabatt.Kode}</Kode>" +
        $"\n            <Belop>{rabatt.Belop}</Belop>" +
        $"\n            <Type>{rabatt.Type}</Type>" +
        $"\n        </Rabatt>";
}
